datafusion/core/src/physical_optimizer/enforce_sorting.rs

```diff
@@ -197,6 +202,41 @@ impl PhysicalOptimizerRule for EnforceSorting {
     }
 }
 
+fn replace_with_partial_sort(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Arc<dyn ExecutionPlan>> {
+    let plan_any = plan.as_any();
+    if let Some(sort_plan) = plan_any.downcast_ref::<SortExec>() {
+        let child = sort_plan.children()[0].clone();
+        if !unbounded_output(&child) {
```

Comment:
Can you add a comment  here about why this operator is only used with unbounded output?

I think it is more generally applicable than for just unbounded cases (it would make any plan more streaming as well as require less memory)

We don't have to do it as part of this PR, but I think we should file a follow on ticket to use this operation more generally

Reply:
I am not completely sure whether it will be safe to expand partial sort without incorporating the `ExternalSorter` used in `SortExec`. Would love to hear your thoughts on this.

Also `SortExec` with unbounded input is already pipeline breaking, I wanted to first gate this change behind unbounded input and improve functionality without regressing the current behaviour.

Opened an issue for expanding PartialSort to other use cases here: https://github.com/apache/arrow-datafusion/issues/9153